File: tasket/data/store.py
from abc import ABC, abstractmethod
from tasket.data import Task, User, Log, TaskFilter, LogFilter
from typing import Optional, List

class UserStore(ABC):
    @abstractmethod
    def get(self, id: str) -> Optional[User]:
        ...

    @abstractmethod
    def update(self, id: str, user: User):
        ...

    @abstractmethod
    def insert(self, user: User):
        ...

    @abstractmethod
    def delete(self, id: str):
        ...


class TaskStore(ABC):
    @abstractmethod
    def get(self, id: str, user_id: str) -> Optional[Task]:
        ...

    @abstractmethod
    def get_by(self, user_id: str, filter: TaskFilter) -> List[Task]:
        ...

    @abstractmethod
    def update(self, id: str, user_id: str, task: Task):
        ...

    @abstractmethod
    def insert(self, task: Task):
        ...


class LogStore(ABC):
    @abstractmethod
    def get_by(self, task_id: str, filter: LogFilter) -> List[Log]:
        ...

File: src/store/store.go
package store

import "github.com/J-Obog/tasket/src/models"

type UserStore interface {
	GetById(id string) (*models.User, error)
	GetByEmail(email string) (*models.User, error)
	Insert(user models.User) error
	Delete(id string) error
}

type TaskStore interface {
	GetById(id string) (*models.Task, error)
	GetByUser(userId string) ([]models.Task, error)
	GetByFilter(userId string, options models.TaskOptions) ([]models.Task, error)
	Update(id string, updatedTask models.UpdatedTask) error
	UpdateStatus(id string, status models.TaskStatus) error
	Insert(task models.Task) error
}

type LogStore interface {
	GetByTask(taskId string) ([]models.Log, error)
	GetByFilter(taskId string, options models.LogOptions) ([]models.Log, error)
	Insert(log models.Log) error
}
